hugo/content/guides/scoping/class-member.md

```diff
@@ -0,0 +1,105 @@
+    )*;
+```
+
+A very important aspect of these chained member calls is the action which rewrites the resulting AST. In this case, it effectively reverses the direction of member call AST nodes. Instead of starting with the first encountered member call and then traversing down to the last, we start with the last and traverse the list of member calls up using the `previous` property. The reason for doing this becomes clear when looking at the scope provider for the Lox language:
+
+```ts
+export class LoxScopeProvider extends DefaultScopeProvider {
+
+    override getScope(context: ReferenceInfo): Scope {
+        // target element of member calls
+        if (context.property === 'element' && isMemberCall(context.container)) {
+            const memberCall = context.container;
+            const previous = memberCall.previous;
+            if (!previous) {
+                return super.getScope(context);
+            }
+            const previousType = inferType(previous);
+            if (isClassType(previousType)) {
+                return this.scopeClassMembers(previousType.literal);
+            }
+            // When the target of our member call isn't a class
+            // Simply return an empty scope
+            return EMPTY_SCOPE;
+        }
+        return super.getScope(context);
+    }
+
+    private scopeClassMembers(classItem: Class): Scope {
+        // Since Lox allows class-inheritance,
+        // we also need to look at all members of possible super classes for scoping
+        const allMembers = getClassChain(classItem).flatMap(e => e.members);
+        return this.createScopeForNodes(allMembers);
+    }
+}
+```
+
+When trying to compute the type of an expression, we are actually only interested in the final piece of the member call. However, to derive the type and scope of the final member call, we have to recursive identify the type of the previous member call. This is done by looking at the member call stored in the `previous` property and inferring its type. See [here](https://github.com/langium/langium-lox/blob/main/src/language-server/type-system/infer.ts) for the full implementation of the type inference system. Note that type inference requires scoping. This leads us to the following invocation graph:
+
+{{<mermaid>}}
+graph LR;
+A(<b>Lox Scope Provider</b><br><i>Member Call</i>) 
+--> B(<b>Type Inference</b><br><i>Class</i>)
+B --> C(<b>Lox Scope Provider</b><br><i>Member Call</i>)
+C --> D(<b>Type Inference</b><br><i>Class</i>)
+D --> E(...)
+E --> F(<b>Lox Scope Provider</b><br><i>Variable Declaration</i>)
+{{</mermaid>}}
+
+We recursively alternate between the scope provider and the type inference system until we arrive at a member call without any previous member call. At this point we resolve the reference using the default lexical scoping which is builtin into Langium. Note that it is recommended to cache type inference information as this naive approach to inference can quickly lead to performance issues.
```

Comment:
Might be also nice to show a quick example, just 2 member calls deep or something, and then connect it with this diagram. Can use one of the examples from above.